core/src/main/java/io/kestra/core/runners/pebble/functions/AbstractFileFunction.java

```diff
@@ -66,8 +65,23 @@ public Object execute(Map<String, Object> args, PebbleTemplate self, EvaluationC
                     fileUri = URI.create(str);
                     namespace = checkAllowedFileAndReturnNamespace(context, fileUri);
                 } else {
-                    namespace = (String) Optional.ofNullable(args.get(NAMESPACE)).orElse(flow.get(NAMESPACE));
-                    fileUri = URI.create(StorageContext.namespaceFilePrefix(namespace) + "/" + str);
+                    if (args.get(NAMESPACE) != null){
+                        namespace = (String) args.get(NAMESPACE);
+                        fileUri = URI.create(StorageContext.namespaceFilePrefix(namespace) + "/" + str);
+                    } else {
+                        namespace = flow.get(NAMESPACE);
+                        fileUri = URI.create(StorageContext.namespaceFilePrefix(namespace) + "/" + str);
+                        String inheritedNamespace = namespace;
+                        URI inheritedFilePath = fileUri;
+                        while (!storageInterface.exists(tenantId, inheritedNamespace, inheritedFilePath) && inheritedNamespace.contains(".")){
+                            inheritedNamespace = inheritedNamespace.substring(0, inheritedNamespace.lastIndexOf('.'));
```

Comment:
Instead of doing substring, better to use the namespace executor which has a method to retrieve all parent namespaces from a namespace